src/cli.rs

```diff
@@ -0,0 +1,821 @@
+) -> Result<String, NysmError> {
+  Ok(match source_format {
+    DataFormat::Json => {
+      let json_value: serde_json::Value = serde_json::from_str(content)?;
+
+      match destination_format {
+        DataFormat::Json => serde_json::to_string_pretty(&json_value)?,
+        DataFormat::Yaml => serde_yaml::to_string(&json_value)?,
+        DataFormat::Text => String::from(content),
+      }
+    }
+    DataFormat::Yaml => {
+      let yaml_value: serde_yaml::Value = serde_yaml::from_str(content)?;
+
+      match destination_format {
+        DataFormat::Json => serde_json::to_string_pretty(&yaml_value)?,
+        DataFormat::Yaml => serde_yaml::to_string(&yaml_value)?,
+        DataFormat::Text => String::from(content),
+      }
+    }
+    DataFormat::Text => String::from(content),
+  })
+}
+
+#[cfg(not(tarpaulin_include))]
+fn pretty_print(content: String, print_format: &DataFormat) -> Result<(), NysmError> {
+  if std::io::stdout().is_terminal() {
+    let language_string = print_format.to_string();
+    let mut printer = PrettyPrinter::new();
+    let _printer = match print_format {
+      DataFormat::Yaml | DataFormat::Json => printer.language(&language_string),
+      _ => &mut printer,
+    };
+
+    #[allow(unused)]
+    #[cfg(not(test))]
+    let _ = _printer
+      .grid(true)
+      .line_numbers(true)
+      .paging_mode(bat::PagingMode::QuitIfOneScreen)
+      .pager("less")
+      .theme("OneHalfDark")
+      .input_from_bytes(content.as_bytes())
+      .print()?;
+  } else {
+    println!("{}", content);
+  }
+
+  Ok(())
+}
+
+fn launch_editor(
+  contents: String,
+  path: tempdir::TempDir,
+  secret_format: &DataFormat,
+  edit_format: &DataFormat,
+) -> Result<Option<String>, NysmError> {
+  let language_string = edit_format.to_string().to_lowercase();
+  let file_path = path.path().join("data").with_extension(language_string);
+
+  let file_contents = reformat_data(&contents, secret_format, edit_format)?;
+  std::fs::write(&file_path, file_contents)?;
+
+  let editor = match std::env::var("EDITOR") {
+    Ok(editor) => editor,
+    Err(_) => String::from("vim"),
+  };
+  let editor_command = format!("{} {}", editor, &file_path.to_string_lossy());
+
+  #[cfg(test)]
+  let editor_command = format!("vim(){{ :; }}; {}", editor_command);
+
+  std::process::Command::new("/usr/bin/env")
+    .arg("sh")
+    .arg("-c")
+    .arg(editor_command)
+    .spawn()
+    .expect("Error: Failed to run editor")
+    .wait()
+    .expect("Error: Editor returned a non-zero status");
+
+  let file_contents: String = std::fs::read_to_string(file_path)?;
+  let json_data = reformat_data(&file_contents, edit_format, secret_format)?;
+
+  if json_data.eq(&contents) {
+    println!("It seems the file hasn't changed, not persisting back to AWS Secrets Manager.");
+
+    Ok(None)
+  } else {
+    Ok(Some(json_data))
+  }
+}
+
+fn temporary_directory() -> std::io::Result<TempDir> {
+  tempdir::TempDir::new("nysm")
+}
```

Comment:
This function is only used once. I don't know if it's planned to be used more times in the future. If not, how necessary is it to abstract it to a separate function like this?

Reply:
The idea is to make where the directory comes from able to be stubbed out. Similar to how the QuerySecrets trait abstracts over the concept of individual client providers in a generic way and can be substituted during testing.